network_test.go

```diff
@@ -0,0 +1,392 @@
+	err := NetworkInterfaces([]NetworkInterface{
+		validStaticNetworkInterface,
+	}).validate(kernelArgsWithIP)
+	assert.Error(t, err, "network interface list with static IP config and IP kernel arg should return validation error")
+}
+
+func TestNetworkInterfacesValidationFails_CNIWithMultiple(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{
+		validCNIInterface,
+		NetworkInterface{
+			StaticConfiguration: &StaticNetworkConfiguration{
+				MacAddress:  "11:22:33:44:55:66",
+				HostDevName: "tap1",
+			},
+		},
+	}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "network interface list with multiple interfaces and CNI configuration should return validation error")
+}
+
+func TestNetworkInterfacesValidationFails_CNIWithKernelArg(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{
+		validCNIInterface,
+	}).validate(kernelArgsWithIP)
+	assert.Error(t, err, "network interface list with CNI config and IP kernel arg should return validation error")
+}
+
+func TestNetworkInterfacesValidationFails_NeitherSpecified(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{{}}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "invalid network config with neither static nor cni configuration did not result in validation error")
+}
+
+func TestNetworkInterfacesValidationFails_BothSpecified(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{{
+		StaticConfiguration: &StaticNetworkConfiguration{
+			MacAddress:  mockMacAddrString,
+			HostDevName: tapName,
+		},
+		CNIConfiguration: validCNIInterface.CNIConfiguration,
+	}}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "invalid network config with both static and cni configuration did not result in validation error")
+}
+
+func TestNetworkMachineCNI(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+	fctesting.RequiresRoot(t)
+
+	cniBinPath := []string{"/opt/cni/bin", testDataBin}
+
+	testCNIDir := filepath.Join(testDataPath, "TestCNI")
+	os.RemoveAll(testCNIDir)
+	defer os.RemoveAll(testCNIDir)
+
+	ipamDataDir := filepath.Join(testCNIDir, "ipam.data")
+
+	cniCacheDir := filepath.Join(testCNIDir, "cni.cache")
+	require.NoError(t,
+		os.MkdirAll(cniCacheDir, 0755),
+		"failed to create cni cache dir")
+
+	cniConfDir := filepath.Join(testCNIDir, "cni.conf")
+	require.NoError(t,
+		os.MkdirAll(cniConfDir, 0755),
+		"failed to create cni conf dir")
+
+	const ifName = "veth0"
+	const networkName = "fcnet"
+
+	cniConf := fmt.Sprintf(`{
+  "cniVersion": "0.3.1",
+  "name": "%s",
+  "plugins": [
+    {
+      "type": "ptp",
+      "ipam": {
+        "type": "host-local",
+        "subnet": "192.168.127.0/24",
+        "resolvConf": "/etc/resolv.conf",
+        "dataDir": "%s"
+      }
+    },
+    {
+      "type": "tc-redirect-tap"
+    }
+  ]
+}`, networkName, ipamDataDir)
+
+	cniConfPath := filepath.Join(cniConfDir, fmt.Sprintf("%s.conflist", networkName))
+	require.NoError(t,
+		ioutil.WriteFile(cniConfPath, []byte(cniConf), 0644),
```

Comment:
Does it have to be writable from the group and others, but not readable?

Reply:
`0644` should result in `-rw-r--r--`, so it's just readable by others here, which is harmless for this configuration file in this test

Reply:
Oh, sorry. Right.